namd/src/colvarproxy_namd.C

```diff
@@ -1539,6 +1551,143 @@ int colvarproxy_namd::smp_colvars_loop()
   return cvm::get_error();
 }
 
+struct cvc_info {
+  bool enabled;
+  bool parent_colvar_total_force_calc;
+  bool parent_colvar_Jacobian;
+  std::shared_ptr<colvar::cvc> cvc_ptr;
+};
+
+// TODO: Too bad! We have to traverse the AST twice!
+void cvc_max_depth(
+  const std::vector<std::shared_ptr<colvar::cvc>>& cvcs,
+  std::unordered_map<std::shared_ptr<colvar::cvc>, int>& cvc_depth_map,
+  int current_level = 0) {
+  for (auto it = cvcs.begin(); it != cvcs.end(); ++it) {
+    auto map_it = cvc_depth_map.find(*it);
+    if (map_it == cvc_depth_map.end()) {
+      cvc_depth_map.insert({*it, current_level});
+    } else {
+      if (map_it->second < current_level) {
+        map_it->second = current_level;
+      }
+    }
+    cvc_max_depth((*it)->children_cvcs(), cvc_depth_map, current_level+1);
+  }
+}
+
+void flatten_all_cvc_tree(
+  std::shared_ptr<colvar::cvc>& parent,
+  std::unordered_map<std::shared_ptr<colvar::cvc>, cvc_info>& cvc_info_map,
+  bool parent_colvar_total_force_calc,
+  bool parent_colvar_Jacobian) {
+  auto children = parent->children_cvcs();
+  for (auto it = children.begin(); it != children.end(); ++it) {
+    flatten_all_cvc_tree(*it, cvc_info_map, parent_colvar_total_force_calc, parent_colvar_Jacobian);
+  }
+  auto it_find = cvc_info_map.find(parent);
+  if (it_find == cvc_info_map.end()) {
+    cvc_info_map.insert({parent, cvc_info{
+      // TODO: Here calc_cvc_values calls cvcs[i]->is_enabled() , which is is_enabled(int f = f_cv_active)
+      //       I know both f_cv_active and f_cvc_active are 0 but are they the same option??
+      parent->is_enabled(colvardeps::features_cvc::f_cvc_active),
+      parent_colvar_total_force_calc, parent_colvar_Jacobian,
+      parent
+    }});
+  }
+}
+
+int colvarproxy_namd::smp_colvars_loop2() {
+  colvarmodule *cv = this->colvars;
+  // Bypass the colvar objects and find all CVC objects
+  std::vector<std::shared_ptr<colvar::cvc>> all_cvcs;
+  std::unordered_map<std::shared_ptr<colvar::cvc>, cvc_info> cvc_info_map;
+  for (auto it = cv->variables_active_smp()->begin();
+       it != cv->variables_active_smp()->end(); ++it) {
+    // TODO: Bad design! What will happen if CVC a is in a "colvar" block
+    // that does not support total_force_calc, but is then reused in
+    // another block that requires total_force_calc even if it supports Jacobian itself???
```

Comment:
I don't see the problem here. Assuming a cvc can have several parents: the colvar that does require a total force can enable it in its children cvcs, even if other parents don't require (or support) it.

Reply:
You are right. I just thought that if a colvar does not require the total force, then it would disable the corresponding feature of the children, but it seems the code do not check the dependency in that way.